### PR TITLE
feat: Dynamic gas EXP

### LINKED ISSUE #206: Compute dynamic gas cost on EXP opcode

### PR DESCRIPTION
**Pull request by Danikim01**
closes #206 
This PR computes the dynamic gas cos for the EXP opcode

### CODE REVIEW

File: src/codegen/operations.rs [hunk 1]
```diff
@@ -131,34 +131,109 @@ fn codegen_exp<'c, 'r>(
...
+    let exponent = stack_pop(context, &ok_block)?;
 
     let result = ok_block
-        .append_operation(ods::math::ipowi(context, lhs, rhs, location).into())
+        .append_operation(ods::math::ipowi(context, base, exponent, location).into())
         .result(0)?
         .into();
 
-    stack_push(context, &ok_block, result)?;
+    // TODO: compute dynamic gas cost depending on the size in bytes of the exponent
```
Comment (ricomateo) [inline src/codegen/operations.rs:154]: We can remove this `TODO` comment

File: src/codegen/operations.rs [hunk 2]
```diff
@@ -131,34 +131,109 @@ fn codegen_exp<'c, 'r>(
...
+            number_of_bytes,
+            constant_value_from_i64(context, &ok_block, 50)?,
+            location,
+        ))
+        .result(0)?
+        .into();
+
+    let total_gas_cost = ok_block
+        .append_operation(arith::addi(
+            constant_value_from_i64(context, &ok_block, 10)?,
```
Comment (ricomateo) [inline src/codegen/operations.rs:207]: Here we can replace the `10` with a constant
```suggestion
            constant_value_from_i64(context, &ok_block, gas_cost::EXP)?,
```

**Review by JulianVentura — COMMENTED** [review]
LGTM

File: src/constants.rs [hunk 3]
```diff
@@ -66,4 +66,27 @@ pub mod gas_cost {
     pub fn log_dynamic_gas_cost(size: u32, topic_count: u32) -> i64 {
         (super::gas_cost::LOG * topic_count as i64) + (8 * size as i64)
     }
+
+    pub fn exponent_byte_size(exponent: u64) -> u32 {
```
Comment (MegaRedHand) [inline src/constants.rs:70]: This should be private
```suggestion
   fn exponent_byte_size(exponent: u64) -> u32 {
```

File: src/constants.rs [hunk 4]
```diff
@@ -66,4 +66,27 @@ pub mod gas_cost {
...
+        ((64 - exponent.leading_zeros()) + 7) / 8
+    }
+
+    pub fn exp_dynamic_cost(exponent: u64) -> u32 {
+        10 + 50 * exponent_byte_size(exponent)
+    }
```
Comment (MegaRedHand) [inline src/constants.rs:76]: These should return a `i64`, so all gas costs are the same.

**Review by MegaRedHand — COMMENTED** [review]
Good work! Left some minor comments.

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**